cpp/src/arrow/compute/exec/tpch_node_test.cc

```diff
@@ -0,0 +1,709 @@
+            if(verify_padding)
+            {
+                int num_chars = ibyte - num_offset;
+                ASSERT_GE(num_chars, 9);
+            }
+        }
+
+        void VerifyStringAndNumber_FixedWidth(
+            const Datum &strings,
+            const Datum &numbers,
+            int byte_width,
+            const char *prefix,
+            bool verify_padding = true)
+        {
+            int64_t length = strings.length();
+            const char *str = reinterpret_cast<const char *>(
+                strings.array()->buffers[1]->data());
+
+            const int32_t *nums = nullptr;
+            if(numbers.kind() != Datum::NONE)
+            {
+                ASSERT_EQ(length, numbers.length());
+                nums = reinterpret_cast<const int32_t *>(
+                    numbers.array()->buffers[1]->data());
+            }
+
+            for(int64_t i = 0; i < length; i++)
+            {
+                const char *row = str + i * byte_width;
+                VerifyStringAndNumber_Single(row, prefix, i, nums, byte_width, verify_padding);
+            }
+        }
+
+        void VerifyStringAndNumber_Varlen(
+            const Datum &strings,
+            const Datum &numbers,
+            const char *prefix,
+            bool verify_padding = true)
+        {
+            int64_t length = strings.length();
+            const int32_t *offsets = reinterpret_cast<const int32_t *>(
+                strings.array()->buffers[1]->data());
+            const char *str = reinterpret_cast<const char *>(
+                strings.array()->buffers[2]->data());
+
+            const int32_t *nums = nullptr;
+            if(numbers.kind() != Datum::NONE)
+            {
+                ASSERT_EQ(length, numbers.length());
+                nums = reinterpret_cast<const int32_t *>(
+                    numbers.array()->buffers[1]->data());
+            }
+
+            for(int64_t i = 0; i < length; i++)
+            {
+                char tmp_str[256] = {};
+                int32_t start = offsets[i];
+                int32_t str_len = offsets[i + 1] - offsets[i];
+                std::memcpy(tmp_str, str + start, str_len);
+                VerifyStringAndNumber_Single(
+                    tmp_str,
+                    prefix,
+                    i,
+                    nums,
+                    sizeof(tmp_str),
+                    verify_padding);
+            }
+        }
+
+        void VerifyVString(const Datum &d, int min_length, int max_length)
+        {
+            int64_t length = d.length();
+            const int32_t *off = reinterpret_cast<const int32_t *>(
+                d.array()->buffers[1]->data());
+            const char *str = reinterpret_cast<const char *>(
+                d.array()->buffers[2]->data());
+            for(int64_t i = 0; i < length; i++)
+            {
+                int32_t start = off[i];
+                int32_t end = off[i + 1];
+                int32_t length = end - start;
```

Comment:
You're shadowing the outer `length` and the outer `i` which makes this a little tricky to read.